agent/lib_monolog.c

```diff
@@ -0,0 +1,387 @@
+ * *
+ * Returns : A new string representing zval; caller must free
+ *
+ */
+static char* nr_monolog_fmt_context_value(zval* zv) {
+  char* val_str = NULL;
+  zval* zv_str = nr_php_zval_alloc();
+
+  ZVAL_DUP(zv_str, zv);
+  convert_to_string(zv_str);
+  val_str = nr_strdup(Z_STRVAL_P(zv_str));
+  nr_php_zval_free(&zv_str);
+
+  return val_str;
+}
+
+/*
+ * Purpose : Format an element of $context array as "key => value" string
+ *
+ * Params  : zval value, pointer to string buffer to store formatted output
+ * and hash key
+ *
+ * Side effect : string buffer is reallocated with each call.
+ *
+ * Returns : ZEND_HASH_APPLY_KEEP to keep iteration
+ *
+ */
+static int nr_monolog_fmt_context_item(zval* value,
+                                       char** strbuf,
+                                       zend_hash_key* hash_key TSRMLS_DC) {
+  NR_UNUSED_TSRMLS;
+  char* key = nr_monolog_fmt_context_key(hash_key);
+  char* val = nr_monolog_fmt_context_value(value);
+
+  char* kv_str = nr_formatf("%s => %s", key, val);
+  nr_free(key);
+  nr_free(val);
+
+  char* sep = nr_strlen(*strbuf) > 1 ? ", " : "";
+  *strbuf = nr_str_append(*strbuf, kv_str, sep);
+  nr_free(kv_str);
+
+  return ZEND_HASH_APPLY_KEEP;
+}
+
+/*
+ * Purpose : Iterate over $context array and format each element
+ *
+ * Params  : string buffer to store formatted output and
+ * Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with Monolog's log context
+ */
+static char* nr_monolog_fmt_context(char* strbuf,
+                                    HashTable* context TSRMLS_DC) {
+  strbuf = nr_str_append(strbuf, "[", "");
+
+  nr_php_zend_hash_zval_apply(context,
+                              (nr_php_zval_apply_t)nr_monolog_fmt_context_item,
+                              (void*)&strbuf TSRMLS_CC);
+
+  return nr_str_append(strbuf, "]", "");
+}
+
+/*
+ * Purpose : Convert $context argument of Monolog\Logger::addRecord to a string
+ *
+ * Params  : # of Monolog\Logger::addRecord arguments, and
+ * Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with Monolog's log context
+ */
+static char* nr_monolog_get_context(size_t argc, NR_EXECUTE_PROTO TSRMLS_DC) {
+  char* context = nr_strdup("");
+  zval* context_arg = NULL;
+
+  if (3 > argc) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: $context not available", __func__);
+    goto return_context;
+  }
+
+  context_arg = nr_php_arg_get(3, NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+
+  if (!nr_php_is_zval_valid_array(context_arg)) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: expected $context be a valid array, got type %d",
+                     __func__, Z_TYPE_P(context_arg));
+    goto return_context;
+  }
+
+  if (0 == nr_php_zend_hash_num_elements(Z_ARRVAL_P(context_arg))) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: $context has no elements", __func__);
+    goto return_context;
+  }
+
+  context = nr_monolog_fmt_context(context, Z_ARRVAL_P(context_arg) TSRMLS_CC);
```

Comment:
Possible mem leak since context had allocated mem [here](https://github.com/newrelic/newrelic-php-agent/pull/489/files#diff-d48bf43b4158fad631ce65b0464841310be10cff57ac0b937061fb9d6c33bdc3R230), but then gets set to another allocated string here.

Reply:
Nice catch - thank you! Good 👁️

Reply:
Upon further inspection this actually is not a memory leak. `nr_monolog_fmt_context` takes the `context` as an argument and then memory pointed to by `context` gets reallocated by `nr_str_append` in `nr_monolog_fmt_context`.